lib/recursive-methods.rb

```diff
@@ -1,49 +1,92 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+  raise ArgumentError if n < 0
+
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return reverse_recursive(s, reversed_string = "", index = -1)
+end
+
+def reverse_recursive(s, reversed_string, index)
+  return reversed_string if index < -s.length
+  reversed_string += s[index]
```

Comment:
`reversed_string += s[index]` creates a new array and  copies all the individual elements over and so is O(n) by itself.